src/document-processor.ts

```diff
@@ -0,0 +1,400 @@
+import {removeNamespaceInitializers} from './passes/remove-namespace-initializers';
+import {removeToplevelUseStrict} from './passes/remove-toplevel-use-strict';
+import {removeUnnecessaryEventListeners} from './passes/remove-unnecessary-waits';
+import {removeWrappingIIFEs} from './passes/remove-wrapping-iife';
+import {rewriteToplevelThis} from './passes/rewrite-toplevel-this';
+import {ConvertedDocumentFilePath, ConvertedDocumentUrl, OriginalDocumentUrl} from './urls/types';
+import {UrlHandler} from './urls/url-handler';
+import {isOriginalDocumentUrlFormat} from './urls/util';
+import {replaceHtmlExtensionIfFound} from './urls/util';
+
+/**
+ * Keep a set of elements to ignore when Recreating HTML contents by adding
+ * code to the top of a program.
+ */
+const generatedElementBlacklist = new Set<string|undefined>([
+  'base',
+  'link',
+  'meta',
+  'script',
+]);
+
+
+
+export type ScanResult =
+    JsModuleScanResult|DeleteFileScanResult|HtmlDocumentScanResult;
+/**
+ * Contains information about how an existing file should be converted to a new
+ * JS Module. Includes a mapping of its new exports.
+ */
+export interface JsModuleScanResult {
+  type: 'js-module';
+  originalUrl: OriginalDocumentUrl;
+  convertedUrl: ConvertedDocumentUrl;
+  convertedFilePath: ConvertedDocumentFilePath;
+  exportMigrationRecords: NamespaceMemberToExport[];
+}
+
+/**
+ * Contains information that an existing file should be deleted during
+ * conversion.
+ */
+export interface DeleteFileScanResult {
+  type: 'delete-file';
+  originalUrl: OriginalDocumentUrl;
+  convertedUrl: undefined;
+  convertedFilePath: undefined;
+}
+
+/**
+ * Contains information that an existing file should be converted as a top-level
+ * HTML file (and not as a new JS module).
+ */
+export interface HtmlDocumentScanResult {
+  type: 'html-document';
+  originalUrl: OriginalDocumentUrl;
+  convertedUrl: ConvertedDocumentUrl;
+  convertedFilePath: ConvertedDocumentFilePath;
+}
+
+
+/**
+ * Converts a Document from Bower to NPM. This supports converting HTML files
+ * to JS Modules (using JavaScript import/export statements) or the more simple
+ * HTML -> HTML conversion.
+ */
+export abstract class DocumentProcessor {
+  protected readonly originalUrl: OriginalDocumentUrl;
+  protected readonly convertedUrl: ConvertedDocumentUrl;
+  protected readonly urlHandler: UrlHandler;
+  protected readonly conversionSettings: ConversionSettings;
+  protected readonly document: Document;
+  protected readonly program: Program;
+  protected readonly convertedHtmlScripts: ReadonlySet<Import>;
+
+  constructor(
+      document: Document, urlHandler: UrlHandler,
+      conversionSettings: ConversionSettings) {
+    this.conversionSettings = conversionSettings;
+    this.urlHandler = urlHandler;
+    this.document = document;
+    this.originalUrl = urlHandler.getDocumentUrl(document);
+    this.convertedUrl = this.convertDocumentUrl(this.originalUrl);
```

Comment:
some background here: The expected behavior of this property is pretty confusing. It is always the "html -> js module converted URL", even if the file is eventually converted as a top-level HTML document. This flaw is even exposed in the return result.

This was only just realized a few weeks ago, since `convertedUrl` is mainly used to create relative URLs, and `convertedFilePath` (which is used to write the conversions to disk) does have the correct file extension.

This doesn't need to be tackled in this PR, but I thought I'd mention it. I'll create a separate issue to track.

Reply:
Noted on the field.